pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -624,14 +624,14 @@ public void run() {
                   _state = State.DISCARDED;
                   break;
                 case DEFAULT:
-                  success = buildSegmentAndReplace();
-                  if (success) {
+                  try {
+                    buildSegmentAndReplace();
```

Comment:
This is not the only place `buildSegmentAndReplace` is being used. It's also being used two times in `goOnlineFromConsuming` method. If we throw exception in buildSegmentAndRepalce, then the segment goes to error state in external view upon receiving helix transition message consuming -> offline. 
The existing behaviour is even worse because segment is not built but external view will move to online state!!
I suggest we proceed with throwing exception in buildSegmentAndReplace, but we try-catch its usage in `goOnlineFromConsuming` and in the catch part, we use `downloadSegmentAndReplace`.
@mcvsubbu @Jackie-Jiang what do you guys think?

Reply:
yes I thought about exactly the same thing. the return null also make the caller alter the state --> if we do not try catch it properly the states wont change.

Ultimately I think we need to rethink when should `_realtimeTableDataManager.addSegmentError` occur, my opinion is that we should make a single state machine function for each functionality, in this case: the only places we catch and process exception should be: `PartitionConsumer` and `goOnlineFromConsuming`, in which both modifies the internal state.

Reply:
adjusted plz kindly take a look

Reply:
The state should be set to DISCARDED if buildSegmentAndReplace fails for whatever reason (It really should not, since another host was able to build). This will allow us to replace the segment by downloading, as a last effort to keep the service going.

What exception are you seeing? Can you paste a stack?

Reply:
First of all, this is not an addSegment error (unless that method in RealtimeTableDataManager has been misnamed). Secondly, this is an "error" that is totally recoverable, since some other host has already completed the segment, and it can be downloaded. So, no "error" should be raised in this condition.

If you can please specify the problem you are trying to solve, we can help.

Reply:
OK, I realized that the addSegmentError is a method that adds all segment errors.  But my question still stands. When we detect a recoverable error, do we still want to add an error? I don't think so. I think we should call addSegmentError() API only when the error is unrecoverable.